tasks_file_string/tasks_string.py

```diff
@@ -0,0 +1,193 @@
+    list_define_popular_time_step_one = [unit for unit in activity_dict.items() if
+                                         int(unit[0]) == int(max_popular_hour_in_day[0]) - 1
+                                         or int(unit[0]) == int(max_popular_hour_in_day[0])
+                                         or int(unit[0]) == int(max_popular_hour_in_day[0]) + 1]
+    list_define_popular_time_step_two = sorted(list_define_popular_time_step_one,
+                                               key=lambda hour_number_records: hour_number_records[1], reverse=True)
+    return list_define_popular_time_step_two
+
+
+def analysis_log_site(source_file_name="task_4_statistics.txt"):
+    """Main function that starts the process of analyzing the log of user activity on the resource"""
+    log_data = open_file(source_file_name)
+    result_file_path_write = check_result_file(source_file_name)
+    all_time_from_log = []  # list of all timestamps of user activity
+    popular_days_of_week = {}  # dictionary by pattern [ip]:[popular days or day of week]
+    ip_usage_list = []  # list of all ip which were active according to the log file
+    for log_line in log_data:
+        list_log_line = [unit.strip() for unit in log_line.split('|')]
+        ip_usage_list.append(list_log_line[0].strip())
+        all_time_from_log.append(((list_log_line[1].strip()).split(":"))[0])
+        creator_dict_ip_day_of_week(list_log_line, popular_days_of_week)
+    ip_usage_dict = Counter(ip_usage_list)  # dictionary where ip it's key and the value of the number of such entries in the log file
+    max_activity_on_day = hour_max_activity_in_day(all_time_from_log)
+    creator_log_analysis_file(popular_days_of_week, ip_usage_dict, max_activity_on_day, result_file_path_write)
+    return f"Result file:{result_file_path_write}"
+
+
+def creator_log_analysis_file(popular_day_of_week, ip_usage_dict, max_activity_on_day, result_file_path_write):
+    """Generates a template for writing to the log analysis file"""
+    for ip in popular_day_of_week:
+        week_day = (str(popular_day_of_week[ip]).lstrip("[")).rstrip("]").upper()
+        ip_count_in = ip_usage_dict[ip]
+        result_string = f"{ip} | most popular days of activity: {week_day} | " \
+                        f"number of visits to the resource: {ip_count_in}\n\n"
+        write_file(result_file_path_write, result_string)
+    if int(max_activity_on_day[0][0]) < int(max_activity_on_day[1][0]):
+        most_popular_hour = f"Most popular times of user activity from {max_activity_on_day[0][0]} " \
+                            f"hour to {max_activity_on_day[1][0]} hour\n"
+        write_file(result_file_path_write, most_popular_hour)
+    most_popular_hour = f"Most popular times of user activity from {max_activity_on_day[1][0]} " \
+                        f"hour to {max_activity_on_day[0][0]} hour\n"
+    write_file(result_file_path_write, most_popular_hour)
```

Comment:
Could you calculate the most popular hour including minutes? So, it will look like: 9:05-10:05, for example.